# Clonal succession after prolonged antiretroviral therapy rejuvenates CD8+ T cell responses against HIV-1

## Article information

*Authors:* Eoghann WHITE<sup>1,2</sup>, Laura PAPAGNO<sup>1</sup>, Assia SAMRI<sup>2</sup>, Kenji SUGATA<sup>3</sup>, Boris HEJBLUM<sup>4</sup>, Amy R. HENRY<sup>5</sup>, Daniel C. ROGAN<sup>6</sup>, Samuel DARKO<sup>5</sup>, Patricia RECORDON-PINSON<sup>7</sup>, Yasmine DUDOIT<sup>8</sup>, Sian LLEWELLYN-LACEY<sup>9</sup>, Lisa A. CHAKRABARTI<sup>10</sup>, Florence BUSEYNE<sup>11</sup>, Stephen A. MIGUELES<sup>6</sup>, David A. PRICE<sup>9,12</sup>, Marie-Aline ANDREOLA<sup>7</sup>, Yorifumi SATOU<sup>3</sup>, Rodolphe THIEBAUT<sup>4,13</sup>, Christine KATLAMA<sup>8</sup>, Brigitte AUTRAN<sup>2</sup>, Daniel C. DOUEK<sup>5</sup> and Victor APPAY<sup>1*</sup>
 
<sup>1</sup> Université de Bordeaux, CNRS, INSERM, ImmunoConcEpT, UMR 5164, Bordeaux, France <br />
<sup>2</sup> Sorbonne Université, INSERM U1135, Centre d’Immunologie et des Maladies Infectieuses (CIMI-Paris), Paris, France <br />
<sup>3</sup> Division of Genomics and Transcriptomics, Joint Research Center for Human Retrovirus Infection, Kumamoto University, Kumamoto, Japan <br />
<sup>4</sup> Université de Bordeaux, INSERM, Bordeaux Population Health Research Centre, U1219, Inria SISTM, Bordeaux, France <br />
<sup>5</sup> Human Immunology Section, Vaccine Research Center, National Institute of Allergy and Infectious Diseases, National Institutes of Health, Bethesda, Maryland, USA <br />
<sup>6</sup> Laboratory of Immunoregulation, National Institute of Allergy and Infectious Diseases, National Institutes of Health, Bethesda, Maryland, USA <br />
<sup>7</sup> Université de Bordeaux, CNRS, Microbiologie Fondamentale et Pathogénicité, UMR 5234, Bordeaux, France <br />
<sup>8</sup> Sorbonne Université, INSERM, Institut Pierre Louis d’Epidémiologie et de Sante Publique, AP-HP, Pitié-Salpêtrière Hospital, Department of Infectious Diseases, Paris, France <br />
<sup>9</sup> Division of Infection and Immunity, Cardiff University School of Medicine, Cardiff, UK <br />
<sup>10</sup> CIVIC Group, Virus and Immunity Unit, Institut Pasteur, Université Paris Cité, CNRS UMR 3569, Paris, France <br />
<sup>11</sup> Unité d'Epidémiologie et Physiopathologie des Virus Oncogènes, Institut Pasteur, Université Paris Cité, CNRS UMR 3569, Paris, France <br />
<sup>12</sup> Systems Immunity Research Institute, Cardiff University School of Medicine, Cardiff, UK <br />
<sup>13</sup> CHU de Bordeaux, Service d’Information Médicale, Bordeaux, France <br />

* Correspondence to Victor Appay (victor.appay@immuconcept.org) 

## Purpose of the GitHub Repository

This GitHub repository contains the scripts and data necessary to reproduce the analysis reported in White et al.
 
## Overview of Repository Structure
	
	├── IMMUNOCO_Directory
	│   ├── data
	│   │   ├── AdditionnalFiles
	│   │   ├── GeneExpression
	│   │   │   ├── processed
	│   │   │   └── raw
	│   │   └── TCR
	│   │       ├── processed
	│   │       │   ├── VDJToolsOutput_TRA
	│   │       │   └── VDJToolsOutput_TRB
	│   │       └── raw
	│   │           ├── MiXCROutput_TRA
	│   │           └── MiXCROutput_TRB
	│   ├── figures
	│   └── scripts
	└── VA02_Directory
	    ├── data
	    │   ├── processed
	    │   └── raw
	    ├── figures
	    └── scripts

The repository is organized into two main directories: one named ImmunocoDirectory, containing scRNAseq data for the Immunoco patients and HIV-2 infected patients, and another named VA02_directory for the VA02 donor. Both directories include subfolders named `data/`, `result/`, and `script/`, which contain raw and processed data when available, the scripts used in this study, and the output figures generated by these scripts. The raw and processed data for the Immunoco donors and HIV-2 infected patients are publicly accessible in the Gene Expression Omnibus (GEO) repository under accession number GSE27065. The raw data for donor VA02 are available in this repository.

## Additional information

- The working environment for each script corresponds to the directory in which the script is located.

- The `clonotypicProcessing_MiXCR.Rmd` and `clonotypicProcessing_VDJTools.Rmd` are designed to operate within a bash environment , utilizing R Markdown to enhance clarity and ease of presentation.

- Due to GitHub's restriction on file sizes exceeding 100 MB, the `matrix.mtx.gz` file for the raw data of the VA02 donor has been split into three separate files. The `NewscRNAseq_Analysis.Rmd` script includes code to merge these split files and reassemble the original `matrix.mtx.gz` file.

